File: sdk-ts/test/manifest.test.js
/**
 * Byte-for-byte parity tests: the TS emitter must reproduce the Rust
 * infra-manifest golden fixture exactly — same JSON text, same
 * content_hash. Regenerate the fixture via the command in README.md.
 */

'use strict';

const { readFileSync } = require('node:fs');
const path = require('node:path');
const { test } = require('node:test');
const assert = require('node:assert/strict');

const { ManifestBuilder, ManifestError, SCHEMA_VERSION } = require('../src/manifest');

const FIXTURE = readFileSync(
  path.join(__dirname, 'fixtures', 'sample.lock.json'),
  'utf8',
).replace(/\r\n/g, '\n');

function buildSample() {
  return new ManifestBuilder('demo')
    .language('ts')
    .bucket('media', { public: false })
    .bucket('public-assets', { public: true })
    .topic('events')
    .topic('eu-events', { region: 'eu-central' })
    .database('main')
    .cron('nightly-rollup', '0 3 * * *')
    .secret('stripe-key')
    .service('api', {
      image: 'registry.example/demo-api:1.4.2',
      ports: { 8080: 8080, 9090: 9090 },
      env: { UPSTREAM: 'https://api.example.internal', RUST_LOG: 'info' },
      buckets: ['media', 'public-assets'],
      topics: ['events', 'eu-events'],
      databases: ['main'],
    })
    .service('worker', {
      image: 'registry.example/demo-worker:1.4.2',
      topics: ['events'],
    });
}

test('ts emitter reproduces rust fixture byte-for-byte', () => {
  const lock = buildSample().finalize();
  assert.equal(lock.toLock(), FIXTURE.replace(/\n$/, ''));
});

test('content_hash matches the rust-stamped hash', () => {
  const rustHash = JSON.parse(FIXTURE).content_hash;
  const lock = buildSample().finalize();
  assert.equal(lock.contentHash, rustHash);
});

test('verify catches post-finalize mutation', () => {
  const lock = buildSample().finalize();
  lock.verify(); // ok
  lock.manifest.services.api.image = 'evil:latest';
  assert.throws(() => lock.verify(), /content_hash mismatch/);
});

test('undeclared resource reference rejected', () => {
  assert.throws(
    () =>
      new ManifestBuilder('demo')
        .service('api', { image: 'nginx', buckets: ['ghost'] }),
    /undeclared/,
  );
});

test('bad names rejected', () => {
  for (const bad of ['UPPER', '9start', 'under_score', '-lead', 'x'.repeat(64), '']) {
    assert.throws(() => new ManifestBuilder(bad), /DNS label/, `app \`${bad}\``);
    assert.throws(
      () => new ManifestBuilder('ok').bucket(bad),
      ManifestError,
      `resource \`${bad}\``,
    );
  }
});

test('duplicate resource rejected; cron needs five fields', () => {
  assert.throws(() => new ManifestBuilder('d').bucket('a').bucket('a'), /duplicate/);
  assert.throws(() => new ManifestBuilder('d').cron('c', '* * *'), /5-field/);
});

test('schema version constant matches rust crate', () => {
  assert.equal(SCHEMA_VERSION, 1);
});

File: sdk-go/manifest_test.go
package bridgebase

import (
	"os"
	"strings"
	"testing"
)

// buildSample mirrors the TS SDK test's buildSample() exactly — same app,
// same declaration order — so the golden fixture pins cross-SDK parity.
func buildSample(t *testing.T) *FinalizedManifest {
	t.Helper()
	b, err := NewBuilder("demo")
	if err != nil {
		t.Fatalf("NewBuilder: %v", err)
	}
	b.Language("ts")
	b.Bucket("media", BucketOpts{Public: false})
	b.Bucket("public-assets", BucketOpts{Public: true})
	b.Topic("events", TopicOpts{})
	eu := "eu-central"
	b.Topic("eu-events", TopicOpts{Region: &eu})
	b.Database("main", "postgres")
	b.Cron("nightly-rollup", "0 3 * * *")
	b.Secret("stripe-key")
	b.Service("api", ServiceSpec{
		Image:     "registry.example/demo-api:1.4.2",
		Ports:     map[int]int{8080: 8080, 9090: 9090},
		Env:       map[string]string{"UPSTREAM": "https://api.example.internal", "RUST_LOG": "info"},
		Buckets:   []string{"media", "public-assets"},
		Topics:    []string{"events", "eu-events"},
		Databases: []string{"main"},
	})
	b.Service("worker", ServiceSpec{
		Image:  "registry.example/demo-worker:1.4.2",
		Topics: []string{"events"},
	})
	fm, err := b.Finalize()
	if err != nil {
		t.Fatalf("Finalize: %v", err)
	}
	return fm
}

// TestGoldenFixtureByteParity is THE parity contract: the Go emitter must
// reproduce sdk-rust/sdk-ts fixtures byte-for-byte, hash included.
func TestGoldenFixtureByteParity(t *testing.T) {
	want, err := os.ReadFile("testdata/sample.lock.json")
	if err != nil {
		t.Fatalf("fixture: %v", err)
	}
	wantText := strings.ReplaceAll(string(want), "\r\n", "\n")

	fm := buildSample(t)
	got := fm.ToLock()

	if got != wantText {
		t.Fatalf("lock text mismatch\n--- got ---\n%s\n--- want ---\n%s", got, wantText)
	}
	if fm.ContentHash() != "4ab6e326f1df50c77348c0681e274e6682604dfc719aacdc36aaaaf7c50c13c7" {
		t.Fatalf("content_hash mismatch: %s", fm.ContentHash())
	}
}

func TestVerifyDetectsTampering(t *testing.T) {
	fm := buildSample(t)
	if err := fm.Verify(); err != nil {
		t.Fatalf("pristine manifest must verify: %v", err)
	}
	fm.obj.app = "tampered"
	if err := fm.Verify(); err == nil {
		t.Fatal("mutated manifest must fail Verify")
	}
}

func TestBuilderErrorsSurfaceAtFinalize(t *testing.T) {
	b, _ := NewBuilder("demo")
	b.Database("main", "mysql") // unsupported engine
	if _, err := b.Finalize(); err == nil {
		t.Fatal("Finalize must return the stored builder error")
	} else if !strings.Contains(err.Error(), "unsupported db engine") {
		t.Fatalf("wrong error: %v", err)
	}
}

func TestUndeclaredResourceRefRejected(t *testing.T) {
	b, _ := NewBuilder("demo")
	b.Service("api", ServiceSpec{Image: "nginx", Databases: []string{"nope"}})
	_, err := b.Finalize()
	if err == nil || !strings.Contains(err.Error(), "undeclared resource") {
		t.Fatalf("want undeclared-resource error, got %v", err)
	}
}

func TestDuplicateAndBadNamesRejected(t *testing.T) {
	b, _ := NewBuilder("demo")
	b.Bucket("x", BucketOpts{})
	b.Bucket("x", BucketOpts{})
	if _, err := b.Finalize(); err == nil || !strings.Contains(err.Error(), "duplicate resource") {
		t.Fatalf("want duplicate error, got %v", err)
	}

	if _, err := NewBuilder("Bad_Name"); err == nil {
		t.Fatal("bad app name must be rejected at construction")
	}

	b2, _ := NewBuilder("demo")
	b2.Bucket("X", BucketOpts{})
	if _, err := b2.Finalize(); err == nil || !strings.Contains(err.Error(), "DNS label") {
		t.Fatalf("want DNS-label error, got %v", err)
	}
}

func TestCronRequiresFiveFields(t *testing.T) {
	b, _ := NewBuilder("demo")
	b.Cron("job", "* * * *") // 4 fields
	if _, err := b.Finalize(); err == nil || !strings.Contains(err.Error(), "5-field cron") {
		t.Fatalf("want cron error, got %v", err)
	}
}

func TestEmptyCollectionsElided(t *testing.T) {
	b, _ := NewBuilder("minimal")
	b.Service("only", ServiceSpec{Image: "nginx:alpine"})
	fm, err := b.Finalize()
	if err != nil {
		t.Fatalf("Finalize: %v", err)
	}
	lock := fm.ToLock()
	for _, banned := range []string{"ports", `"env"`, "buckets", "topics", "databases"} {
		if strings.Contains(lock, banned) {
			t.Fatalf("empty field %q must be elided:\n%s", banned, lock)
		}
	}
	if !strings.Contains(lock, `"language": "go"`) == false && b.language != nil {
		t.Fatal("unexpected language emission")
	}
}
